# Analyzing borrowers’ risk of defaulting

The goal of the project is to prepare a report for a bank’s loan division. We’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

The report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

## Step 1. Open the data file and have a look at the general information. 

In [1]:
# importing libraries and reading the data file 

#! pip install nltk -U -q
import pandas as pd
import numpy as np
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english') 

/Users/pavellugovoy/opt/anaconda3/envs/practicum/lib/python3.7/site-packages/sklearn/feature_extraction/image.py:167: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  dtype=np.int):
/Users/pavellugovoy/opt/anaconda3/envs/practicum/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Depr

In [2]:
# reading the file
credit_scoring = pd.read_csv('')

Looking at the general information

In [3]:
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [5]:
credit_scoring.tail(10)

children  days_employed  dob_years            education  education_id  \
21515         1    -467.685130         28  secondary education             1   
21516         0    -914.391429         42    bachelor's degree             0   
21517         0    -404.679034         42    bachelor's degree             0   
21518         0  373995.710838         59  SECONDARY EDUCATION             1   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21515            married                 0      F    employee     1   
21516            married                 0      F    business     0   
21517  civil partnership                 1      F    business     0   
21518            married                 0      F     retiree     0   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                     purpose  
21515     17517.812          to become educated  
21516     51649.244    purchase of my own house  
21517     28489.529           buying my own car  
21518     24618.344           purchase of a car  
21519     18551.846  buy commercial real estate  
21520     35966.698        housing transactions  
21521     24959.969           purchase of a car  
21522     14347.610                    property  
21523     39054.888           buying my own car  
21524     13127.587                to buy a car

In [6]:
credit_scoring.sample()

children  days_employed  dob_years            education  education_id  \
276         0   -8631.590122         64  secondary education             1   

    family_status  family_status_id gender income_type  debt  total_income  \
276       married                 0      F    employee     0     25730.322   

                purpose  
276  housing renovation

In [7]:
credit_scoring.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

### Conclusion

The first look at the data let us conclude that: 

1. We have information not only on customers' marital status, number of children, and the customers' debt default, but also on the customers' employment period, age, education, gender, income type, income and purpose of loan. This information may help us to achieve the goal of our research. 

2. At the same time we see that not everything is fine with our data: 
- the columns containing the customers' employment period and income have null values, 
- some values reflecting the employment period are negative, so they are incorrect, 
- the values reflecting the days of employment and income are of "float" type, that seems to be inappropriate (if we measure time by "days", we do not need normally to compare the different periods by "part" of the days),
- some values in the 'education' column have the same meaning but they are written in a different way, so it may cause problems when we analyse them.

We still don't know:  

1. which types of marital status are in the table,  

2. if everything is ok with the values of column of 'family_status', 'family_status_id', as well as  'children', 'days_employed','dob_years', 'income_type', 'total_income'.


So we need to proceed with:

1.  further analyzing the data to find out if there are problems that we don't know from the first look, 

2.  understanding which problems we should handle in order to accomplish our analysis

2.  prepocessing the data so that we could make our research in the most effecient way. 

## Step 2. Data preprocessing

### Processing missing values

Looking for missing values within category values

In [8]:
credit_scoring['income_type'].value_counts()

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
unemployed                         2
entrepreneur                       2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [9]:
credit_scoring['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [10]:
credit_scoring['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

In [11]:
credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [12]:
credit_scoring['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

In [13]:
credit_scoring['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [14]:
credit_scoring['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

In [15]:
credit_scoring['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
buying property for renting out             653
housing transactions                        653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

Looking for missing values within quantative values

In [16]:
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [17]:
# filling in missing values in 'children'
credit_scoring['children'] = credit_scoring['children'].replace([20],2)
credit_scoring['children'] = credit_scoring['children'].replace([-1],1)

# checking the results
credit_scoring['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Looking for other missing values in the column 'days_employed'

In [18]:
credit_scoring['days_employed'].value_counts()

-8437.673028      1
-3507.818775      1
 354500.415854    1
-769.717438       1
-3963.590317      1
                 ..
-1099.957609      1
-209.984794       1
 398099.392433    1
-1271.038880      1
-1984.507589      1
Name: days_employed, Length: 19351, dtype: int64

In [19]:
credit_scoring['days_employed'].unique()

array([-8437.67302776, -4024.80375385, -5623.42261023, ...,
       -2113.3468877 , -3112.4817052 , -1984.50758853])

Checking values in the column 'dob_years'

In [20]:
# looking for missing values in the column 'dob_years'

credit_scoring['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Looking for missing values in the column 'total_income'

In [21]:
credit_scoring['total_income'].value_counts()

17312.717    2
42413.096    2
31791.384    2
21005.772    1
18591.443    1
            ..
27020.895    1
23686.835    1
9606.294     1
28156.762    1
13127.587    1
Name: total_income, Length: 19348, dtype: int64

Checking if the rows with the missing values in 'total_income" are the same rows as in the column 'days_employed'

In [22]:
credit_scoring[credit_scoring['total_income'].isnull()].count()

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

In [23]:
credit_scoring[credit_scoring['days_employed'].isnull()].count()

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

These are actually the same rows.

We need to look at the percentiles in order to have a general understanding of the distribution of values. We'll use it to decide which value to choose (the mean or the median) to fill in the missing values. 

In [24]:
credit_scoring.groupby(['income_type']).describe(percentiles = [0.3, 0.50, 0.6,0.7,0.98,1])

children                                          \
                               count      mean       std  min  30%  50%  60%   
income_type                                                                    
business                      5085.0  0.524287  0.756333  0.0  0.0  0.0  0.0   
civil servant                 1459.0  0.603153  0.848312  0.0  0.0  0.0  1.0   
employee                     11119.0  0.576850  0.803621  0.0  0.0  0.0  1.0   
entrepreneur                     2.0  0.000000  0.000000  0.0  0.0  0.0  0.0   
paternity / maternity leave      1.0  2.000000       NaN  2.0  2.0  2.0  2.0   
retiree                       3856.0  0.094139  0.335851  0.0  0.0  0.0  0.0   
student                          1.0  0.000000       NaN  0.0  0.0  0.0  0.0   
unemployed                       2.0  0.500000  0.707107  0.0  0.3  0.5  0.6   

                                             ...  total_income                \
                             70%   98% 100%  ...          mean           std   
income_type                                  ...                               
business                     1.0  2.00  5.0  ...  32386.793835  20876.975327   
civil servant                1.0  3.00  5.0  ...  27343.729582  15500.602149   
employee                     1.0  3.00  5.0  ...  25820.841683  14611.602368   
entrepreneur                 0.0  0.00  0.0  ...  79866.103000           NaN   
paternity / maternity leave  2.0  2.00  2.0  ...   8612.661000           NaN   
retiree                      0.0  1.00  4.0  ...  21940.394503  12839.512522   
student                      0.0  0.00  0.0  ...  15712.260000           NaN   
unemployed                   0.7  0.98  1.0  ...  21014.360500  16152.074628   

                                                                            \
                                   min         30%         50%         60%   
income_type                                                                  
business                      4592.450  21533.7278  27577.2720  31377.9126   
civil servant                 4672.012  18103.2305  24071.6695  27304.0204   
employee                      3418.824  17672.0858  22815.1035  25667.2640   
entrepreneur                 79866.103  79866.1030  79866.1030  79866.1030   
paternity / maternity leave   8612.661   8612.6610   8612.6610   8612.6610   
retiree                       3306.762  14374.4360  18962.3180  21776.7636   
student                      15712.260  15712.2600  15712.2600  15712.2600   
unemployed                    9593.119  16445.8639  21014.3605  23298.6088   

                                                                              
                                    70%          98%        100%         max  
income_type                                                                   
business                     36131.0938  86498.82512  362496.645  362496.645  
civil servant                31245.4774  70195.73746  145672.235  145672.235  
employee                     29121.2158  63341.77652  276204.162  276204.162  
entrepreneur                 79866.1030  79866.10300   79866.103   79866.103  
paternity / maternity leave   8612.6610   8612.66100    8612.661    8612.661  
retiree                      25239.7862  58494.80192  117616.523  117616.523  
student                      15712.2600  15712.26000   15712.260   15712.260  
unemployed                   25582.8571  31978.75234   32435.602   32435.602  

[8 rows x 77 columns]

Filling the missing values in 'total_income' with median 

In [25]:
credit_scoring['total_income'] = (credit_scoring
                                  .groupby(['income_type'])['total_income']
                                  .transform(
                                      lambda grp: grp.fillna(np.mean(grp)))
                                 )

In [26]:
# checking the results

credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [27]:
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

In [28]:
credit_scoring.tail(10)

children  days_employed  dob_years            education  education_id  \
21515         1    -467.685130         28  secondary education             1   
21516         0    -914.391429         42    bachelor's degree             0   
21517         0    -404.679034         42    bachelor's degree             0   
21518         0  373995.710838         59  SECONDARY EDUCATION             1   
21519         1   -2351.431934         37      graduate degree             4   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21515            married                 0      F    employee     1   
21516            married                 0      F    business     0   
21517  civil partnership                 1      F    business     0   
21518            married                 0      F     retiree     0   
21519           divorced                 3      M    employee     0   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                     purpose  
21515     17517.812          to become educated  
21516     51649.244    purchase of my own house  
21517     28489.529           buying my own car  
21518     24618.344           purchase of a car  
21519     18551.846  buy commercial real estate  
21520     35966.698        housing transactions  
21521     24959.969           purchase of a car  
21522     14347.610                    property  
21523     39054.888           buying my own car  
21524     13127.587                to buy a car

In [29]:
credit_scoring.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.479721   63046.497661     43.293380      0.817236   
std        0.755528  140827.311974     12.574584      0.548138   
min        0.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   21525.000000  
mean           0.972544      0.080883   26783.346520  
std            1.420324      0.272661   15665.112650  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   17247.708000  
50%            0.000000      0.000000   24309.013000  
75%            1.000000      0.000000   32386.793835  
max            4.000000      1.000000  362496.645000

### Conclusion

### Data type replacement

In [30]:
#changing data type of the values in the column 'total' from to integer
credit_scoring['total_income'] = credit_scoring['total_income'].astype('int')

#checking results
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int64  
 11  purpose           21525 non-null  object 
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


### Conclusion

The column 'total_income' contains the values of float type. For our research the float type is excessive, we do not need to know every cent to make a conclusion, while the short numbers look easier and better.  

### Processing duplicates

In [31]:
# finding the number of duplicates 
credit_scoring.duplicated().sum()

54

In [32]:
# looking once again at values of column 'education'
credit_scoring['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [33]:
# applying lower() method to the values of 'education' so that more potential duplicates coud be detected
credit_scoring['education'] = credit_scoring['education'].str.lower()

In [34]:
# checking the results
credit_scoring['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [35]:
# looking at the duplicates once again
credit_scoring.duplicated().sum()

71

In [36]:
# dropping out the duplicates
credit_scoring = credit_scoring.drop_duplicates().reset_index(drop=True)

In [37]:
# checking the results
credit_scoring.duplicated().sum()

0

### Conclusion

We found initially 54 duplicates. But  the values in the column 'education' are written differently but having the same meaning. This may have impact on the number of detected values. We've used str.lower() to equalize the values of this column.  After that we've found 71 duplicated rows. So that means that these rows are filled with exactly the same data including 'total_income', 'dob_years', 'days_employed', 'purpose' and other columns. It may be caused by human error: an operator when non-intentially copied these rows two or more times extracting them 'by hand' from one data base to another. These rows relate to the same customers and therefore we cannot and we should not merge them to keep some data calculated. Such duplicates make data dirtier (we would analyse the same cases as different), so we need to drop them out. We have dropped them out using drop_duplicates().    

### Categorizing Data

In [38]:
credit_scoring['family_status'].value_counts()

married              12339
civil partnership     4151
unmarried             2810
divorced              1195
widow / widower        959
Name: family_status, dtype: int64

In [39]:
credit_scoring['total_income'].describe(percentiles = [0.05, 0.20, 0.50, 0.80, 0.95])

count     21454.000000
mean      26788.610189
std       15689.695960
min        3306.000000
5%        10330.850000
20%       15765.600000
50%       24301.500000
80%       34338.600000
95%       53050.450000
max      362496.000000
Name: total_income, dtype: float64

In [40]:
# setting up variables to describe the income level milestones

low_level = credit_scoring['total_income'].quantile(0.05).astype('int')
low_middle_level = credit_scoring['total_income'].quantile(0.20).astype('int')
middle_level = credit_scoring['total_income'].quantile(0.5).astype('int')
high_level = credit_scoring['total_income'].quantile(0.80).astype('int')
upper_high_level = credit_scoring['total_income'].quantile(0.95).astype('int')

In [41]:
# setting up a function enabling us to categorize the data

def income_group(income):
    
    """the function returns the income_group according to the income value using following rules:
     'lowest_income' for income <= low_level
     'low_income' for low_level < income <= low_middle_level
     'low_middle_income' for low_middle_level < income <= middle_level
     'middle_income' for middle_level < income <= high_level
     'high_income' for high_level < income <= upper_high_level
     'upper_high_income' for income > upper_high_income"""
    
    if income <= low_level:
        return 'lowest_income'
    elif low_level < income <= low_middle_level:
        return 'low_income'
    elif low_middle_level < income <= middle_level:
        return 'low_middle_income'
    elif middle_level < income <= high_level:
        return 'middle_income'
    elif high_level < income <= upper_high_level:
        return 'high_income'
    else:
        return 'upper_high_income'

In [42]:
# checking the function 

income_group(93000)

'upper_high_income'

In [43]:
income_group(8000)

'lowest_income'

In [44]:
income_group(13000)

'low_income'

In [45]:
income_group(18000)

'low_middle_income'

In [46]:
income_group(25000)

'middle_income'

In [47]:
income_group(36000)

'high_income'

In [48]:
#creating a new column 

credit_scoring['income_group'] = credit_scoring['total_income'].apply(income_group)

In [49]:
#checking the results

credit_scoring.head(10)
print(credit_scoring['income_group'].value_counts())

low_middle_income    6436
middle_income        6436
high_income          3218
low_income           3218
lowest_income        1073
upper_high_income    1073
Name: income_group, dtype: int64


In [50]:
# categorizing by purpose group 

credit_scoring['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            768
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

There're different ways of spelling the same words. We need to standardtize them for further processing. 

In [51]:
# finding out the stemms corresponding to the words in question
words = ['university', 'car', 'education', 'housing', 'estate', 'property', 'wedding']
for word in words: 
    print(english_stemmer.stem(word))

univers
car
educ
hous
estat
properti
wed


In [52]:
# setting up the function to process the values of the rows of the 'purpose' column 

def purpose_group(purpose):
    
    """ the funcion returns purpose group according to the purpose value using following rules: 
    if the purpose contains a word with the stem 'wed' the function returns 'wedding'
    if the purpose contains a word with the stem 'hous' or 'properti' or 'estat' the function returns 'real_estate'
    if the purpose contains a word with the stem 'car' the function returns 'car'
    if the purpose contains a word with the stem 'educ' or "univers' the function returns 'education"""
    try: 
        
        for word in purpose.split(' '): 
        
            stemmed_word = english_stemmer.stem(word)
            if stemmed_word == 'wed':
                return 'wedding'
            elif stemmed_word == 'hous' or stemmed_word == 'properti' or stemmed_word == 'estat':
                return 'real_estate'
            elif stemmed_word == 'car':
                return 'car'
            elif stemmed_word == 'educ' or stemmed_word == 'univers':
                return 'education'
        
    except: 
        
        stemmed_word = english_stemmer.stem(purpose)
        if stemmed_word == 'wed':
            return 'wedding'
        elif stemmed_word == 'hous' or stemmed_word == 'properti' or stemmed_word == 'estat':
            return 'real_estate'
        elif stemmed_word == 'car':
            return 'car'
        elif stemmed_word == 'educ' or stemmed_word == 'univers':
            return 'education'

In [53]:
# applying the function purpose_group(purpose)
credit_scoring['purpose_group'] = credit_scoring['purpose'].apply(purpose_group)

In [54]:
# checking the results
credit_scoring.head(10)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  secondary education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  secondary education             1   
8         2   -6929.865299         35    bachelor's degree             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   
5  civil partnership                 1      M    business     0         40922   
6            married                 0      F    business     0         38484   
7            married                 0      M    employee     0         21731   
8  civil partnership                 1      F    employee     0         15337   
9            married                 0      M    employee     0         23108   

                               purpose       income_group purpose_group  
0                purchase of the house        high_income   real_estate  
1                         car purchase  low_middle_income           car  
2                purchase of the house  low_middle_income   real_estate  
3              supplementary education        high_income     education  
4                    to have a wedding      middle_income       wedding  
5                purchase of the house        high_income   real_estate  
6                 housing transactions        high_income   real_estate  
7                            education  low_middle_income     education  
8                     having a wedding         low_income       wedding  
9  purchase of the house for my family  low_middle_income   real_estate

#### Conclusion

Among the values we need to anylise we have two columns with a vast list of values: the first contains information of customers' income ('total_income') and the other one contains information on the loan purpose ('purpose'). 

1. The values of the total_income column are that different for "natural" reasons: different people have different income, so that's why there is so wide range. 

But it is not very convinient to analyse such data: we need to compare many values, so we need some categorization of these values.  We'have divided these values into 6 groups according to the following rules: 
- the group with the lowest income is the group of customers with the income lower than rest 95% of others, 
- the the group with low income is the group of those having income higher than than the lowset income group but lower than  rest 80% of others,
- the group with low middle income is the group of those having income higher than than the low income group but lower than  rest 50% of others,
- the group with middle income is the group of those having income higher than than the low middle income group but lower than rest 20% of others,
- the group with high income is the group of those having income higher than middle income group ibut lower than the rest 5% of others,
- the group with upper high income is the group of those having income higher than high income group.

To do this we have used quantile() method to determine the milestones and we have written a function returning names of each category. We have put new values to  a new column "income_group'.

2. The values of purpose group may differ for several reasons. Maybe, it is explained by the content of questionaires customers fill in when they get a loan (ther are many options to choose). But due to the "style" of the definitions used to describe the purpose the most probable reason is that  customers define the purpose on their own, so they write what they wish and the definitions vary, this is "natural". We have divided the purposes of loan into 4 groups: wedding needs, real estate needs, car purchase needs, and education needs.  To do this we have first looked at values, then with SnowballStemmer method we have defined the relevant "stemms" for key words describing the said reasons. And finally we have written a function returning names to fill in the values a new column with the said grouped purposes.

There is also a column with values relating to marital status ('family_status'). These values are also very clos to each other ('divorced' is close to 'unmarried' and "widow/widower", as they are all single). But in fact they materially differ, because the legal status and civil rights and obligations of each of the type vary, so it may have impact on their loan payment discipline. That's why we decided to leave these data "as is" 

## Step 3. Answer these questions

### Is there a relation between having kids and repaying a loan on time?

In [55]:
# calculating absolute numbers
with_children = len(credit_scoring[credit_scoring.children >= 1])
without_children = len(credit_scoring[credit_scoring.children == 0])
with_children_defaulted = len(credit_scoring[(credit_scoring.children >= 1) & (credit_scoring.debt == 1)])
without_children_defaulted = len(credit_scoring[(credit_scoring.children == 0) & (credit_scoring.debt == 1)])

In [56]:
with_children_defaulted

678

In [57]:
without_children_defaulted

1063

In [58]:
# calculating ratios
with_children_default_ratio = with_children_defaulted / with_children 
without_children_default_ratio = without_children_defaulted / without_children 
with_children_to_without_children_ratio = with_children_default_ratio / without_children_default_ratio 

In [59]:
with_children_default_ratio

0.09208203178052424

In [60]:
without_children_default_ratio

0.07543822297920659

In [61]:
with_children_to_without_children_ratio

1.2206283253239576

In [62]:
# aggregating data into pivot table
scoring_pivot_kids  = credit_scoring.pivot_table(index='children', values='debt', aggfunc='sum')
scoring_pivot_kids ['ratio'] = credit_scoring.pivot_table(index='children', values='debt', aggfunc='sum') / credit_scoring.pivot_table(index='children', values='debt', aggfunc='count')
scoring_pivot_kids.sort_values('ratio',  ascending=False)

debt     ratio
children                
4            4  0.097561
2          202  0.094925
1          445  0.091658
3           27  0.081818
0         1063  0.075438
5            0  0.000000

#### Conclusion

We found the ratio of customers having kids having defaulted to all customers having kids and the same value for those non-having kids. And those who have kids defaulted  for 122.06 % more often than those having one or more kid. 
We have then looked deeper into data and found some interesting things. The general rule is that "the more kids have the customer the worse borrower she or he is", but we have found out that people with 5 kids had never defaulted, as well as people with 3 kids were better clients than those with 1,2 and 4 childs. 

### Is there a relation between marital status and repaying a loan on time?

In [63]:
# aggregating data into a pivot table

scoring_pivot_family = credit_scoring.pivot_table(index='family_status', values='debt', aggfunc='sum')
scoring_pivot_family['ratio'] = credit_scoring.pivot_table(index='family_status', values='debt', aggfunc='sum') / credit_scoring.pivot_table(index='family_status', values='debt', aggfunc='count')
                                 
scoring_pivot_family.sort_values('ratio',  ascending=False)


debt     ratio
family_status                    
unmarried           274  0.097509
civil partnership   388  0.093471
married             931  0.075452
divorced             85  0.071130
widow / widower      63  0.065693

#### Conclusion

We have calculated for each marital status group ratio of number of customers having defaulted to all customers of grop and have found that unmarried persons and customers being in civil partnership defaulted more than others, widows and widowers defaulted the less, and the middle of the statistics are married and divorced. 

### Is there a relation between income level and repaying a loan on time?

In [64]:
# aggregating data into a pivot table

scoring_pivot_income = credit_scoring.pivot_table(index='income_group', values='debt', aggfunc='sum')
scoring_pivot_income['ratio'] = credit_scoring.pivot_table(index='income_group', values='debt', aggfunc='sum') / credit_scoring.pivot_table(index='income_group', values='debt', aggfunc='count')
                                 
scoring_pivot_income.sort_values('ratio',  ascending=False)

debt     ratio
income_group                     
low_middle_income   558  0.086700
low_income          273  0.084835
middle_income       539  0.083748
high_income         225  0.069919
upper_high_income    75  0.069897
lowest_income        71  0.066170

#### Conclusion

We have calculated for each income group ratio of number of customers having defaulted to all customers of group and have found that the best borrowers were customers with lowest income level. At the second place w are the customers with high and upper high income with similar default ratio as the lowest income customers. Middle income customers paid worse, low income customers paid a little bit worse than middle income level customers but better than low_middle_income customers which paid the worst.  

### How do different loan purposes affect on-time repayment of the loan?

In [65]:
# aggregating data into a pivot table

scoring_pivot_purpose = credit_scoring.pivot_table(index='purpose_group', values='debt', aggfunc='sum')
scoring_pivot_purpose['ratio'] = credit_scoring.pivot_table(index='purpose_group', values='debt', aggfunc='sum') / credit_scoring.pivot_table(index='purpose_group', values='debt', aggfunc='count')
                                 
scoring_pivot_purpose.sort_values('ratio',  ascending=False)

debt     ratio
purpose_group                
car             403  0.093590
education       370  0.092200
wedding         186  0.080034
real_estate     782  0.072334

#### Conclusion

We have calculated for each loan purpose customers group the ratio of castomers having defaulted to all customers of the group and we have found that the worst clients were those who took loans for car purchase needs. They are followed by those who took loans for education. The cusomers who borrowed money for wedding purposed were a little bit better than the first three, but the best customers were those took loans for real estate needs. 

## Step 4. General conclusion


I. INTRODUCTION 

The goal of this analysis was to determine if a customer’s marital status, number of children, income level, loan purposes have an impact on whether they will default on a loan. 

In order to perform this analysis we used data on customers’ credit worthiness regarding previous period. We assume that  if nothing is materially changed in human behavior due to economic and/or legal environment the conclusions we made based on this analysis may be used for future customers' banking scoring purposes. 

The data contains the following information regarding each of the customers: 

1. number of customer's children;
2. number of days of customer's employment period;
3. customer's education; 
4. customer's marital status;
5. customer's gender;
6. customer's age;
7. customer's income type;
8. customer's debt default; 
9. customer's income; 
10. purpose of customer's loan.

We preanalysed the data and found out that there were several issues with the values of the data and the some data need to be preprocessed before the analysis. 


II. PREPROCESSING 

A. Missing values. 

We discovered missing values in the columns of the table related to the information on number of customer's children, number of days of customer's employment period, customer's age, customer's income. Here is the way how we processed these missing values.  


1. Missing values in information on related to number of days of customer's employment period and customer's income. 
The number of rows with missing values in the columns 'days_employed' (employment period) and 'total_income' (customer's income) is 2174, and these are the same rows. So, probably, these are the customers which just have not provided information on their employment conditions (income level and employment period).  Such rows make 10% of all rows, so it is quite significant, so normally we need to keep them if it is possible. 

1.1. Missing values in 'days_employed'.

Thanks to describe() method and other methods we applied to the values of this columnm we found out that these values were strange: the mean of all values is  63046 days, i.e. 172 years.  It is not just possibe in real world. At the same time we found out that there were many negative values. So if we had replaced them with positive ones, the mean would have been more higher. We don't know the reason, but this data is either very damaged or represents some other information (not employment period). It was not possible to restore them without having additional information. As we did not need it for our research we decided to leave these values "as is". 


1.2. Missing values in "total_income". 
Unlike the employment period information on customers' total income is important for us, because we further used it in our research. 
The question was which value or values to choose to fill in these one. We assumed that the customers with different income type ('retiree', 'emlpoyee', 'business' and so on) had different average and median income level. Hence we analysed income level of customers divided into different categories. We found that there were no significant outliers in these groups and we decided therefore to fill the missing values with a average value belonging to the relevant income type category.  
 

2. Missing values in 'children'.

We some strange values in the information related to the information on the quantity of customer's children: 76 customers with 20 children and 47 customers with a negative number of children. 20 children in a family is a rare case.  Within the customers of the bank, for example, there were only 9 customers with 5 children and 41 with with 4. So this information is most probably false, and it is probably a result of a typo, and it should be "2" instead of "20". Taking into account that the number of the rows with missing values was not signifacnt compare to the total number of customers, so if we had made a mistake it would not have affected dramaticcaly the data. Hence  we changed the value of these rows to "2" - the most probable" number in this case. The same logic we applied to the  the negative number: it was oubviously false, it was a typo. The number of inccorect values was not significant, so we changed this value to "1".

3. Missing values in 'dob_years'.

In the column 'dob_years' there were 101 rows with zero values. It means that there were 101 customers information on the age thereof was misssing, because among the customers of the bank there was no kid. Normally we woud have filled in these missing values with the median/mean values for each type of the customers (as we did with the missing values on the income level), but as this data was not necessary for our analysis, we left this data unchanged.  

B. Changing data type. 

The column 'total_income' (customer's income) and 'dob_years' (age) contained the values of float type. We don't need the age information for our analysism, so we didn't change it. For our research the float type for values of customers' income was excessive, as we did not need to know every cent to make a conclusion, while the short numbers look easier and better. Hence we changed the data type of these values from float to integer.

C. Processing duplicates. 

We found initially 54 duplicates. But the values in the column 'education' were written differently but had the same meaning. This had impact on the number of detected duplicated values. After we equalized the values of this column we found 71 duplicated rows. So that means that these rows are filled with exactly the same data including 'total_income', 'dob_years', 'days_employed', 'purpose' and other columns. It might be caused by human error: an operator non-intentially copied these rows two or more times extracting them 'by hand' from one data base to another. These rows relate to the same customers and therefore we could not and we should not merge them to keep some data calculated. Such duplicates would have made data dirtier (we would have analysed the same cases as different), so we dropped them out using drop_duplicates() method. 

D. Categorizing Data. 

Among the values we need to anylize we had two columns with a vast list of values: the first related to the information of customers' income ('total_income') and the other to the information on the loan purpose ('purpose'). 

1. Customer's income. 
The values of the total_income column are that different for "natural" reasons: different people have different income, so that's why there is so wide range. 

But it is not very convinient to analyse such data: we need to compare many values, so we need some categorization of these values.  We divided these values into 6 groups according to the following rules: 
- the group with the lowest income is the group of customers with the income lower than rest 95% of others, 
- the the group with low income is the group of those having income higher than than the lowset income group but lower than  rest 80% of others,
- the group with low middle income is the group of those having income higher than than the low income group but lower than  rest 50% of others,
- the group with middle income is the group of those having income higher than than the low middle income group but lower than rest 20% of others,
- the group with high income is the group of those having income higher than middle income group ibut lower than the rest 5% of others,
- the group with upper high income is the group of those having income higher than high income group.

To do this we have used quantile() method to determine the milestones and we have written a function returning names of each category. We have put new values to  a new column "income_group'.

2. Purpose category. 
The values of 'purpose' might differ for several reasons. It may be explained by the content of questionaires customers fill in when they get a loan (ther are many options to choose). But due to the "style" of the definitions used to describe the purpose the most probable reason is that  customers define the purpose on their own, so they write what they wish and the definitions vary, this is "natural". We have divided the purposes of loan into 4 groups: 
- wedding needs, 
- real estate needs, 
- car purchase needs, and 
- education needs.  

To do this we first looked at values, then with SnowballStemmer method we  defined the relevant "stemms" for key words describing the said reasons. And finally we wrote a function returning names to fill in the values a new column with the said grouped purposes.

There is also a column with values relating to marital status ('family_status'). These values are also very close to each other ('divorced' is close to 'unmarried' and "widow/widower", as they are all single). But in fact they materially differ, because the legal status and civil rights and obligations of each of the type vary, so it may have impact on their loan payment discipline. That's why we decided to leave these data "as is". 


III. RESEARCH 

A. Having kids and paying loan in time. 

We found the ratio of customers having kids having defaulted to all customers having kids ('Having kids default ratio') and the same value for those non-having kids ('Non-having kids default ratio'). 

Here are the results: 

1. Having kids default ratio is 0,92
2. Non-having kids default ratio is 0,75

It means that those who have no kids were more reliable than those having kids.

At the same time we found an exception to this rule: all of  9 customers with 5 children have never defaulted the payments. 

B. Marital status and payment default.

We have calculated for each marital status group ratio of number of customers having defaulted to all customers of grop and have found that unmarried persons and customers being in civil partnership defaulted more than others, widows and widowers defaulted the less, and the middle of the statistics are married and divorced.

Here are the exact ratios found in descending order: 

1. unmarried         0.097509
2. civilpartnership  0.093471
3. married           0.075452 
4. divorced          0.071130
5. widow / widower   0.065693
                   


C. Income level and payment default. 

We calculated for each income group ratio of number of customers having defaulted to all customers of such a group and found that the best borrowers were customers with lowest income level! That was a little surprise. At the second place are the customers with high and upper high income with similar default ratio as the lowest income customers. Middle income customers paid worse, low income customers paid a little bit worse than middle income level customers but better than low_middle_income customers which paid the worst. 

Here are the exact ratio numbers in descending order: 
                
1. low middle income     0.086700
2. low income            0.084835
3. middle income         0.083748
4. high income           0.069919
5. upper high_income     0.069897
6. lowest income         0.066170


D. Loan purposes and payment default.

We calculated for each loan purpose customers group the ratio of customers having defaulted to all customers of the group.  We  found that the worst clients were those who took loans for car purchase needs. They are followed by those who took loans for education. The customers who borrowed money for wedding purpose were a little bit better than the first three, but the best customers were those who took loans for real estate needs. 

Here are the exact results: 
           
1. car               0.093590
2. education         0.092200
3. wedding           0.080034
4. real_estate       0.072334



IV. ADDITIONAL REMARKS 

The results we found may be used to evaluate the client reliability. But we always should take into account that along with the general line there is a place for an "exception". We see it, for example, in the case of customers with kids and without kids, and the case of lowest income borrowers, which are were the best clients, while low income clients perform their payments worse than those who earn more. 




### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [x]  an explanation of which missing value types were detected;
- [x]  explanation for the possible causes of missing values;
- [x]  an explanation of how the blanks are filled;
- [x]  replaced the real data type with an integer;
- [x]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [x]  an explanation of which method is used to find and remove duplicates;
- [x]  description of the possible reasons for the appearance of duplicates in the data;
- [x]  data is categorized;
- [x]  an explanation of the principle of data categorization;
- [x]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [x]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [x]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [x]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [x]  conclusions are present on each stage;
- [x]  a general conclusion is made.